File: tests/pyspm/matlab/prepareOrthView.m
function prepareOrthView(mat, dim)
% set structure for Display and draw a first overlay
global strParam
strParam = struct('n', 0, 'bb',[],'Space',eye(4),'centre',[0 0 0],'mode',1,...
    'area',[0 0 1 1],'premul',eye(4),'hld',1,'modeDispl',[0, 0, 0]);
strParam.Space = spm_matrix([0 0 0  0 pi -pi/2])*strParam.Space;    

% get bounding box and resolution
if isempty(strParam.bb) 
     strParam.maxBB = maxbb(mat, dim); 
     strParam.bb = strParam.maxBB;    
end
resolution(mat);

% Draw at initial location, center of bounding box
mmcentre     = mean(strParam.Space*[strParam.maxBB';1 1],2)';
strParam.centre    = mmcentre(1:3);
% Display modes: [Background+Stat+ROIs, Background+Stat, Background+ROIs]
strParam.modeDispl = [0 0 1];

return;

function bb = maxbb(mat, dim)
global strParam
mn = [Inf Inf Inf];
mx = -mn;
premul = strParam.Space \ strParam.premul;

d = dim;
corners = [
    1    1    1    1
    1    1    d(3) 1
    1    d(2) 1    1
    1    d(2) d(3) 1
    d(1) 1    1    1
    d(1) 1    d(3) 1
    d(1) d(2) 1    1
    d(1) d(2) d(3) 1
    ]';
XYZ = mat(1:3, :) * corners;

XYZ = premul(1:3, :) * [XYZ; ones(1, size(XYZ, 2))];
bb = [
    min(XYZ, [], 2)'
    max(XYZ, [], 2)'
    ];

mx = max([bb ; mx]);
mn = min([bb ; mn]);
bb = [mn ; mx];
return;


function resolution(mat)
global strParam
resDefault = 1; % Default minimum resolution 1mm
res = min([resDefault,sqrt(sum((mat(1:3,1:3)).^2))]);
res      = res / mean(svd(strParam.Space(1:3,1:3)));
Mat      = diag([res res res 1]);
strParam.Space = strParam.Space * Mat;
strParam.bb    = strParam.bb / res;
return;

File: opennft/prepare_orth_view.py
import numpy as np
from rtspm import spm_matrix, spm_imatrix


def prepare_orth_view(mat, dim):
    # set structure for Display and draw a first overlay
    str_param = {'n': 0, 'bb': [], 'space': np.eye(4, 4), 'centre': np.zeros((1, 3)), 'mode': 1,
                 'area': np.array([0, 0, 1, 1]), 'premul': np.eye(4, 4), 'hld': 1, 'mode_displ': np.zeros((1, 3))}

    temp = np.array([0, 0, 0, 0, np.pi, -np.pi / 2])
    str_param['space'] = spm_matrix.spm_matrix(temp) @ str_param['space']

    # get bounding box and resolution
    if len(str_param['bb']) == 0:
        str_param['max_bb'] = max_bb(mat, dim, str_param['space'], str_param['premul'])
        str_param['bb'] = str_param['max_bb']

    str_param['space'], str_param['bb'] = resolution(mat, str_param['space'], str_param['bb'])

    # Draw at initial location, center of bounding box
    temp = np.vstack((str_param['max_bb'].T, [1, 1]))
    mmcentre = np.mean(str_param['space'] @ temp, 1)
    str_param['centre'] = mmcentre[0:3]
    # Display modes: [Background + Stat + ROIs, Background + Stat, Background + ROIs]
    str_param['mode_displ'] = np.array([0, 0, 1])

    return str_param


def max_bb(mat, dim, space, premul):

    mn = np.array([np.inf] * 3, ndmin=2)
    mx = -mn
    premul = np.linalg.solve(space, premul)
    bb, vx = get_bbox(mat, dim, premul)
    mx = np.vstack((bb, mx)).max(0)
    mn = np.vstack((bb, mx)).min(0)
    bb = np.vstack((mn, mx))

    return bb


def get_bbox(mat, dim, premul):
    p = spm_imatrix(mat)
    vx = p[6:9]

    corners = np.array([
        [1, 1, 1, 1],
        [1, 1, dim[2], 1],
        [1, dim[1], 1, 1],
        [1, dim[1], dim[2], 1],
        [dim[0], 1, 1, 1],
        [dim[0], 1, dim[2], 1],
        [dim[0], dim[1], 1, 1],
        [dim[0], dim[1], dim[2], 1],

    ]).T

    xyz = mat[0:3, :] @ corners

    xyz = premul[0:3, :] @ np.vstack((xyz, np.ones((1, xyz.shape[1]))))

    bb = np.array([
        np.min(xyz, axis=1).T,
        np.max(xyz, axis=1).T
    ])

    return bb, vx


def resolution(mat, space, bb):
    res_default = 1

    temp = (np.sum((mat[0:3, 0:3]) ** 2, axis=0) ** .5)
    res = np.min(np.hstack((res_default, temp)))

    u, s, v = np.linalg.svd(space[0:3, 0:3])
    temp = np.mean(s)
    res = res / temp

    mat = np.diag([res, res, res, 1])

    space = space @ mat
    bb = bb / res

    return space, bb
